stan/math/rev/functor/reduce_sum.hpp

```diff
@@ -66,10 +78,17 @@ struct reduce_sum_impl<ReduceFunction, require_var_t<ReturnType>, ReturnType,
     recursive_reducer(recursive_reducer& other, tbb::split)
         : num_vars_per_term_(other.num_vars_per_term_),
           num_vars_shared_terms_(other.num_vars_shared_terms_),
-          sliced_partials_(other.sliced_partials_),
+          varis_(other.varis_),
+          partials_(other.partials_),
           vmapped_(other.vmapped_),
           msgs_(other.msgs_),
-          args_tuple_(other.args_tuple_) {}
+          args_tuple_(other.args_tuple_),
+          main_thread_(other.main_thread_),
+          args_adjoints_holder_(
+              std::make_unique<double[]>(num_vars_shared_terms_)),
```

Comment:
Well here the unique_ptr is constructed pointing to the memory address `num_vars_shared_terms_`, and in the first constructor it points to `nullptr`.

I don't understand how `args_adjoints_` works in this copied version. Like isn't it going to be a length `num_vars_shared_terms_` Map pointing at memory address `num_vars_shared_terms_`?

Reply:
So there are two cases:

1. the root worker node is mapping directly into the partials array of the final output
2. the split off workers have their own memory space where they accumulate the adjoints of the partials (as the workers get to evaluate multiple work chunks not just one)

So for case 1 I don't need to allocate anything such that it points nowhere (nullptr). For the case 2, we have to allocate an array which holds the adjoints of the shared terms. The unique_ptr is there to store a reference to that and make sure we wipe out the memory upon destruction. The Eigen Map then needs to be setup with a pointer from the unique_ptr and the size is num_vars_shared_terms (one adjoint per var).

Does it now click? (or do I have a crack)?

Reply:
> For the case 2, we have to allocate an array which holds the adjoints of the shared terms.

Oh I didn't realize that's what `unique_ptr<double[]>(N)` did. Will stare some more.